sharedbuffers/mapped_struct.py

```diff
@@ -1719,6 +1742,420 @@ def __str__(self):
+            (eint < 128 and (self.bitrep_hi & (1 << (eint - 64))) != 0))
+
+    @cython.locals(lo=cython.Py_ssize_t, hi=cython.Py_ssize_t, mid=cython.Py_ssize_t,
+        dcode=cython.char, objlen=cython.longlong, itemsize=cython.uchar,
+        offset=cython.Py_ssize_t, h1=cython.ulonglong, h2=cython.ulonglong)
+    def __contains__(self, elem):
+        if self.objlist is None:
+            return self._contains_compressed(elem)
+
+        lo = 0
+        hi = len(self.objlist)
+        dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+
+        if cython.compiled:
+            if dcode not in ('t', 'T'):
+                return self._search_key(elem, dcode, offset, lo, hi, hi // 2, True) < hi
+
+        h2 = _stable_hash(elem)
+        while lo < hi:
+            mid = lo + ((hi - lo) >> 1)
+            val = self.objlist._c_getitem(mid, dcode, objlen, itemsize, offset, _struct, None)
+            h1 = _stable_hash(val)
+            if h1 == h2:
+                # Equal hashes - Look in both directions for a match, without binary search
+                if val == elem:
+                    return True
+                while mid > 0:
+                    val = self.objlist._c_getitem(mid, dcode, objlen, itemsize, offset, _struct, None)
+                    h1 = _stable_hash(val)
+                    if h1 != h2:
+                        break
+                    elif val == elem:
+                        return True
+                    mid -= 1
+
+                mid = lo + ((hi - lo) >> 1)
+                while mid < hi:
+                    val = self.objlist._c_getitem(mid, dcode, objlen, itemsize, offset, _struct, None)
+                    h1 = _stable_hash(val)
+                    if h1 != h2:
+                        break
+                    elif val == elem:
+                        return True
+                    mid += 1
+                return False
+            elif h1 < h2:
+                lo = mid + 1
+            else:
+                hi = mid
+        return False
+
+    @cython.locals(i=cython.Py_ssize_t, dcode=cython.char, objlen=cython.longlong,
+        itemsize=cython.uchar, offset=cython.Py_ssize_t)
+    def __iter__(self):
+        if self.objlist is not None:
+            dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+            for i in xrange(len(self)):
+                yield self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None)
+        else:
+            if self.bitrep_lo:
+                for i in xrange(64):
+                    if self.bitrep_lo & (1 << i):
+                        yield i
+                    elif not self.bitrep_lo >> i:
+                        break
+
+            if self.bitrep_hi:
+                for i in xrange(64):
+                    if self.bitrep_hi & (1 << i):
+                        yield i + 64
+                    elif not self.bitrep_hi >> i:
+                        break
+
+    def union(self, *seqs):
```

Comment:
@klaussfreire It is ok to maintain these operations slow?

Reply:
On the one side, most of these operations have to create a new set, so they're screwed in that sense.

On the other side, it's possible we can implement optimized construction of that new set at least. We will use them, so we want them as fast as practical.

For instance, in the case of bitmap sets on both sides, most operations can be extremely efficient. Most other cases are tougher though.

Reply:
At the very least, a fast-call version of `frozenset(self)` that avoids having to iterate self through the pure-python interface could be added.